src/main/java/org/opensearch/knn/index/codec/KNN990Codec/NativeEngines990KnnVectorsReader.java

```diff
@@ -350,4 +287,76 @@ private static List<String> getVectorCacheKeysFromSegmentReaderState(SegmentRead
 
         return cacheKeys;
     }
+
+    private void loadMemoryOptimizedSearcherIfRequired() {
+        if (vectorSearchers != null) {
+            return;
+        }
+
+        synchronized (this) {
+            if (vectorSearchers != null) {
+                return;
+            }
+
+            // We need sufficient memory space for this table as it will be queried for every single search.
+            // Hence, having larger space to approximate a perfect hash here.
+            final Map<String, VectorSearcher> vectorSearcherPerField = new HashMap<>(
+                RESERVE_TWICE_SPACE * segmentReadState.fieldInfos.size(),
+                SUFFICIENT_LOAD_FACTOR
+            );
+
+            try {
+                for (FieldInfo fieldInfo : segmentReadState.fieldInfos) {
+                    final IOSupplier<VectorSearcher> searcherSupplier = getIndexFileNameIfMemoryOptimizedSearchSupported(fieldInfo);
+                    if (searcherSupplier != null) {
+                        final VectorSearcher searcher = searcherSupplier.get();
+                        if (searcher != null) {
+                            // It's supported. There can be a case where a certain index type underlying is not yet supported while
+                            // KNNEngine
+                            // itself supports memory optimized searching.
+                            vectorSearcherPerField.put(fieldInfo.getName(), searcher);
+                        }
+                    }
+                }
+
+                vectorSearchers = vectorSearcherPerField;
+            } catch (Exception e) {
+                // Close opened searchers first, then suppress
+                try {
+                    IOUtils.closeWhileHandlingException(vectorSearcherPerField.values());
+                } catch (Exception closeException) {
+                    log.error(closeException.getMessage(), closeException);
+                }
+                throw new RuntimeException(e);
+            }
+        }
+    }
+
+    private IOSupplier<VectorSearcher> getIndexFileNameIfMemoryOptimizedSearchSupported(final FieldInfo fieldInfo) {
```

Comment:
nit: name of this method looks wrong - I would expect it to return a file name not a vector searcher. Can you fix?